src/net/protocol_agnostic/listener.rs

```diff
@@ -140,37 +191,132 @@ impl PaListener {
+        match self.inner {
+            PaIncomingInner::Tcp { ref mut incoming, ref mut processing } => {
+                loop {
+                    match incoming.poll().map_err(AcceptError::TcpAccept)? {
+                        Async::Ready(Some((stream, addr))) => {
+                            processing.push({
+                                let header = [0u8; 8];
+
+                                tokio_io::io::read_exact(stream, header)
+                                .map_err(AcceptError::TcpReadHeader)
+                                .and_then(move |(stream, header)| {
+                                    match header {
+                                        ECHO_REQ => {
+                                            tcp_respond_with_addr(stream, addr)
+                                            .map(|_stream| None)
+                                            .map_err(AcceptError::TcpRespond)
+                                            .into_boxed()
+                                        },
+                                        CRUST_TCP_INIT => {
+                                            future::ok(Some((stream, addr)))
+                                            .into_boxed()
+                                        },
+                                        _ => {
+                                            future::err(AcceptError::InvalidTcpHeader)
+                                            .into_boxed()
+                                        },
+                                    }
+                                })
+                                .into_boxed()
+                            });
+                        }
+                        Async::Ready(None) |
+                        Async::NotReady => break,
+                    }
+                }
+                loop {
+                    match processing.poll()? {
+                        Async::Ready(Some(Some((stream, addr)))) => {
+                            let stream = PaStream::Tcp(stream);
+                            let addr = PaAddr::Tcp(addr);
+                            return Ok(Async::Ready(Some((stream, addr))));
+                        }
+                        Async::Ready(Some(None)) => (),
+                        Async::Ready(None) |
+                        Async::NotReady => break,
                     }
-                    Async::Ready(None) => Ok(Async::Ready(None)),
-                    Async::NotReady => Ok(Async::NotReady),
                 }
+                Ok(Async::NotReady)
             }
-            PaIncoming::Utp(ref mut incoming) => {
-                match incoming.poll()? {
+            PaIncomingInner::Utp { ref mut incoming, ref mut raw_rx, ref mut processing } => {
+                loop {
+                    match raw_rx.poll().void_unwrap() {
+                        Async::Ready(Some(raw_channel)) => {
+                            processing.push({
```

Comment:
This body of this match arm could be extracted to separate function. That would make code a little more readable, maintainable, testable, .etc